packages/core/upload/server/controllers/admin-api.js

```diff
@@ -173,15 +174,23 @@ module.exports = {
       request: { files: { files } = {} },
     } = ctx;
 
-    if (id && (_.isEmpty(files) || files.size === 0)) {
-      return this.updateFileInfo(ctx);
+    try {
+      if (id && (_.isEmpty(files) || files.size === 0)) {
+        return this.updateFileInfo(ctx);
+      }
+
+      if (_.isEmpty(files) || files.size === 0) {
+        throw new ApplicationError('Files are empty');
+      }
+
+      await (id ? this.replaceFile : this.uploadFiles)(ctx);
+    } finally {
+      if (Array.isArray(files)) {
+        files.map(file => rimraf(file.path, () => {}));
+      } else if (files.path) {
+        rimraf(files.path, () => {});
+      }
```

Comment:
It may be better to do that into a middleware, at the end at all requests. Wdyt ? Checking if there is `ctx.request.files`

Reply:
I think I would do this in the upload or replace service instead. 
You can use fs-extra to do the delete without installing rimraf as a dependency by the way

Reply:
But what if a middleware wants to use the files afterwards? Then they would not be available anymore.
Thus my proposition to delete them inside a middleware a the correct place (first place, after the next()).

I need rimraf anyway somewhere else to delete a whole folder (doesn't work well natively in Node 12).
But you're right I'll change for `unlink` that can be awaited, it's prettier

Reply:
@alexandrebodin What do you think?

Reply:
No, exactly, I'm proposing to change that to a middleware. However I have to check if I can control the middleware order enough